### PR TITLE
[REVIEW] Add support for fill() on dictionary columns

### PR DESCRIPTION
**Pull request by davidwendt**
This will add specialization code for support the `fill()` API on dictionary columns.
The API accepts a scalar and a column to fill into. 
The dictionary specialization will require the scalar-type match the dictonary's keys' type.
The overall process will be to add the new scalar value to the keyset, update the indices, and then do a normal fill operation on just the indices.
This will require a new dictionary function to retrieve the index for a specific key. This value is then used to fill the target indices column.

### CODE REVIEW

File: cpp/src/dictionary/search.cu [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    std::unique_ptr<numeric_scalar<int32_t>> operator()( dictionary_column_view const& input, scalar const& key,
+                                                         rmm::mr::device_memory_resource* mr, cudaStream_t stream ) const
+    {
+        auto result = std::make_unique<numeric_scalar<int32_t>>( 0, false, stream, mr );
+        if( input.size()==0 )
+            return result;
+        CUDF_EXPECTS( input.keys().type()==key.type(), "search key type must match dictionary keys type" );
+        auto keys_view = column_device_view::create(input.keys(),stream);
+        auto find_key = static_cast<experimental::scalar_type_t<Element> const&>(key);
+        auto iter = thrust::copy_if( rmm::exec_policy(stream)->on(stream),
```
Comment (devavret) [inline cpp/src/dictionary/search.cu:65]: Why not a thrust::find? I imagine it'd be faster than a copy_if.

Comment (davidwendt) [reply cpp/src/dictionary/search.cu:65]: So `thrust::find()` returns an iterator which then I can diff with to get the index but then I need to copy the integer into device memory via numeric_scalar. The `thrust::copy_if()` does both in one kernel -- find and copy.

**Comment by davidwendt** [comment]
> Instead of making a new search specifically for dictionary columns, you can re-use the functionality in `cudf/search.hpp`. You need to pass the dictionay column's keys column to it and a 1 element column for search value.

The search APIs in `cudf/search.hpp` are `upper_bound()`, `lower_bound()` and two `contains()` functions. The `contains()` ones return bools which are not helpful. And the two `_bound()` functions will always return a valid result.
https://github.com/rapidsai/cudf/blob/627ebc51cc48188ea06c801bc1924e72060789ed/cpp/include/cudf/search.hpp#L30-L31
Whereas I want to know if the key does not exist and return a `numeric_scalar` with `is_valid()==false` in that case.

File: cpp/include/cudf/dictionary/detail/search.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * If the key does not exist in the dictionary the returned scalar will have `is_valid()==false`
+ * 
+ * @throw cudf::logic_error if `key.type() != dictionary.keys().type()`
+ * 
+ * @param dictionary The dictionary to search for the key.
+ * @param key The value to search for in the dictionary keyset.
+ * @return index value of the key within the dictionary
+ */
+std::unique_ptr<numeric_scalar<int32_t>> get_index( dictionary_column_view const& dictionary, scalar const& key,
```
Comment (vuule) [inline cpp/include/cudf/dictionary/detail/search.hpp:38]: Is the dictionary index always int32_t?

Comment (davidwendt) [reply cpp/include/cudf/dictionary/detail/search.hpp:38]: Yes,

File: cpp/src/dictionary/set_keys.cu [hunk 3]
```diff
@@ -53,9 +53,8 @@ struct dispatch_compute_indices
             [d_dictionary] __device__ (size_type idx) {
                 if( d_dictionary.is_null(idx) )
                     return Element{};
-                column_device_view d_indices = d_dictionary.child(0);
                 column_device_view d_keys = d_dictionary.child(1);
-                auto index = d_indices.element<int32_t>(idx + d_dictionary.offset());
+                size_type index = static_cast<size_type>(d_dictionary.element<dictionary32>(idx));
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/dictionary/set_keys.cu:57]: No.

File: cpp/src/dictionary/search.cu [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        if( d_keys.is_null(idx) )
+            return false;
+        return d_key == d_keys.element<Element>(idx);
+    }
+};
+
+/**
+ * @brief Find index of a given key within a dictionary's keys column.
+ * 
+ * The index is the position within the keys column where the given key (scalar) is found.
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/dictionary/search.cu:48]: Why is that better?

File: cpp/src/dictionary/search.cu (hunk 1 above)
Comment (davidwendt) [reply cpp/src/dictionary/search.cu:65]: I expect to find the key most of the time.

File: cpp/src/dictionary/search.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/dictionary/search.cu:48]: It's much harder for the caller to miss the fact that the key was not found.

[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/dictionary/search.cu:48]: I guess I don't agree with that. I'd rather reserve nullptr for something more unexpected. I feel an invalid scalar would be a more appropriate, modern return. Like `std::find()` and others return an iterator that you check for `end()`.

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/dictionary/search.cu:48]: >>I'd rather reserve nullptr for something more unexpected.
We presumably throw an exception in such cases.

I'm not strongly arguing for returning null. As long as it's consistent with other APIs, I'm fine with both options.

File: cpp/include/cudf/dictionary/detail/search.hpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace cudf
+{
+namespace dictionary
+{
+namespace detail
+{
+
+/**
+ * @brief Return the index value for a given key.
```
Comment (harrism) [inline cpp/include/cudf/dictionary/detail/search.hpp:28]: Use @copydoc?

**Pull request merged**